mesonbuild/backend/ninjabackend.py

```diff
@@ -2623,7 +2623,7 @@ def generate_genlist_for_target(self, genlist: build.GeneratedList, target: buil
                 args = [x.replace('@DEPFILE@', depfile) for x in base_args]
             args = [x.replace("@INPUT@", infilename).replace('@OUTPUT@', sole_output)
                     for x in args]
-            args = self.replace_outputs(args, self.get_target_private_dir(target), outfilelist)
+            args = self.replace_outputs(args, self.get_target_private_dir(target), outfiles)
```

Comment:
outfilelist is the list of all outputs for genlist.

outfiles is the list of outputs for the current file, with the target private dir included -- which is now being doubled, as far as I can tell.

Reply:
> which is now being doubled, as far as I can tell.

I just saw the error, it makes the test fails when target is in a sub-directory.

As far as I understand I should `replace_outputs` with `outfiles` before the private part is added, but I need to pass the paths with the private part to ` NinjaBuildElement`

Reply:
Save two variables? :)

Reply:
that's what I did locally, but I didn't managed to the issue in unit tests yet (using public api), so I have second though whether this is the right fix or not, especially since `outfilelist` is supposed to be updated to remove the generated outputs ([here](https://github.com/mesonbuild/meson/blob/43b80e02ce0e87dfcf069111e62ad8eff4435d6e/mesonbuild/backend/ninjabackend.py#L2629))

the following change triggers the issue for me

```diff
--- a/mesonbuild/modules/_qt.py
+++ b/mesonbuild/modules/_qt.py
@@ -463,7 +463,7 @@ class QtBaseModule(ExtensionModule):
         # depfile arguments (defaults to <output-name>.d)
         DEPFILE_ARGS: T.List[str] = ['--output-dep-file'] if self._moc_supports_depfiles else []
 
-        arguments = kwargs['extra_args'] + DEPFILE_ARGS + inc + compile_args + ['@INPUT@', '-o', '@OUTPUT@']
+        arguments = kwargs['extra_args'] + DEPFILE_ARGS + inc + compile_args + ['@INPUT@', '-o', '@OUTPUT0@']
         preserve_path_from = os.path.join(state.source_root, state.subdir) if kwargs['preserve_paths'] else None
         if kwargs['headers']:
             moc_gen = build.Generator(
```

Reply:
> but I didn't managed to the issue in unit tests yet

PEBKAC, nevermind

I added a test case that reproduce the issue, I removed the usage of  `outfilelist` completely as it makes more sens to rely on `genlist.get_outputs_for(curfile)` to know the list of output files for a given entry

Reply:
ping

Reply:
ping